### ISSUE TITLE
More tutorials needed

### ISSUE THREAD

**Issue by IvanSanchez** [Leaflet/Leaflet#4239]
The [tutorials](http://leafletjs.com/examples.html) page needs some love. Leaflet has a lot of features, but there are tutorials for just a bunch of them.

We should spend some time converting some of the pages in the `debug/` directory into tutorials/examples.

A non-comprehensive list:

* [ ] Complete newbie's guide to How To Use a Leaflet Plugin
* [ ] Image overlays (prettier version of `image-overlay.html`)
* [ ] Map bounds (with `bounceAtBounds`)
* [ ] Layer opacity
* [ ] WMS (with special attention to Leaflet/WMS nomenclature about the term "layer")
* [ ] Fractional zoom controls (prettier version of `zoomdelta.html`)
* [ ] setView, setZoom, fitBounds and flyTo (prettier version of `zoompan.html`)
* [ ] `L.CRS.Simple`
* [ ] Custom CRSs (using, for example, a [map of the galaxy](https://en.wikipedia.org/wiki/Milky_Way#/media/File:Artist%27s_impression_of_the_Milky_Way_%28updated_-_annotated%29.jpg) with galactic CRS)
* [ ] A primer on how to create Leaflet plugins (`L.Class`, `L.Class.include` vs `L.Class.extend`, etc)

**Comment by mourner** [comment]
@IvanSanchez fantastic CRS tutorial, glad to see this being finally addressed! 
Also see #3271 for previous discussion on tutorials.

**Comment by Kayoku** [comment]
Hi everyone.
I'm just discovering Leaflet, then i did [this](http://leafletjs.com/examples/quick-start.html) "Quick Start Guide". This is maybe stupid but, when you need to make sure that the map container has a defined height with : 
`#mapclass { height: 180px; }`
that's not working. Because we did this before : 
`<div id="mapid"></div>`
Just rename "mapclass" to "mapid" ? (For beginner, it's maybe more clear)
What do you think ?

**Comment by mourner** [comment]
Hmm, yeah, that's a bug in the tutorial — well noted! Would you want to fix this yourself and make a pull request? It's as easy as clicking "edit" on this file https://github.com/Leaflet/Leaflet/blob/gh-pages/examples/quick-start.md, making a change right in the browser and saving.

**Comment by Kayoku** [comment]
Ok thanks. Correction [#4317](https://github.com/Leaflet/Leaflet/pull/4317) open !

**Comment by skorasaurus** [comment]
Sorry for making that mistake and thank you for correcting it @Kayoku :)

**Comment by iH8** [comment]
The thing is having the most lovable API. You'll need an example for every piece of Leaflet. A couple of tutorials won't suffice. Running/forkable examples. After that tutorials will be a breeze. Promise, willing to resolve. So we can reference those while writing tutorials. Every layer needs to be covered. You can't expect people to grasp tutorials but you can state them the reference. If we'll have running examples for everything running all the versions i'll stfu for ever.

**Comment by mourner** [comment]
Replying to a comment about tutorials that was apparently removed by the author, I think eventually we would want, in addition to the current narrated tutorials, build a set of live examples for every feature, just like [Mapbox does for mapbox.js](https://www.mapbox.com/mapbox.js/example/v1.0.0/). But meanwhile, the latter is a great resource, because 95% of the examples are not Mapbox-specific and can be used with Leaflet.

**Comment by ughitsaaron** [comment]
I hope it's alright for me to throw in and say I'd be happy to help with some of this docs/tutorial writing in the near future. I've been using leaflet a lot lately for a side project I'm working on and would love to help contribute to it eventually.

**Comment by IvanSanchez** [comment]
@ughitsaaron Absolutely! I'd love to see some pull requests for more examples/demos!

If you're working on something, just leave a comment about what your example is about, and submit a pull request (to the `gh-pages` branch) when done.

If you need any guidance on how to create a new example and edit the files needed, just ask.

**Comment by smanohar3** [comment]
Hi Everyone,

I would like to contribute to this. Can you let me know on how to proceed. I'm a newbie on Github!

**Comment by ghybs** [comment]
Hi @smanohar3,

Thank you for your message!
Sorry for not getting back to you earlier.

You can have a look at [the other tutorials](https://github.com/Leaflet/Leaflet/tree/master/docs/examples) as a starting point.

Instead of plain HTML files, Leaflet website pages (including the tutorials) are written in [GitHub Markdown](https://github.github.com/gfm/) with [Jekyll](https://jekyllrb.com/) templating (pretty standard for GitHub-hosted pages)
But you can still use plain HTML within the Markdown file.

As for contributing:
- If you are familiar with git, you can [fork](https://guides.github.com/activities/forking/) the repository, create a new branch and commit your modifications into that branch. Then push to your remote fork and [create a Pull Request](https://help.github.com/articles/creating-a-pull-request-from-a-fork/). You can also perform all these tasks directly on GitHub website.
- If your modification is very simple (typically modifying a single file), you can directly [edit the file on GitHub](https://help.github.com/articles/editing-files-in-another-user-s-repository/), which will directly fork, create a branch, commit and offer you to create a Pull Request.

Feel free to ask for help.
There are also plenty resources on the web to help you using GitHub and contributing in general.

We would be very happy to receive your first contribution, so please persevere during the learning process! (we know it is not easy at the beginning…) :smiley:

**Comment by brunoclima** [comment]
Just noted that the tutorials page needs a refresh. There are many broken examples like [this](https://leafletjs.com/examples/wms/wms.html).

**Comment by vncntcltt** [comment]
Hi there,

I have some spare time at the moment. 
I can help with some of these tutorials, starting with investigating the boken ones.
Regarding the WMS example, it seems like the [boundlessgeo demo service](https://demo.boundlessgeo.com/) is no longer accessible.

**Comment by vncntcltt** [comment]
I can write new tutorials for marker.cluster usage and Vuejs integration without using [Vue2Leaflet](https://github.com/vue-leaflet/Vue2Leaflet) (and possibly for React integration as well).

These tutorials would be more or less based on this demo: https://github.com/vncntcltt/airport-map

@IvanSanchez, @mourner, @ghybs: do you think it would be useful ?

**Comment by johnd0e** [comment]
> These tutorials would be more or less based on this demo: https://github.com/vncntcltt/airport-map

Broken atm (https://vncntcltt.github.io/airport-map/)
```
Failed to load resource: the server responded with a status of 404 ()
vncntcltt.github.io/:1 Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0```

**Comment by vncntcltt** [comment]
> > These tutorials would be more or less based on this demo: https://github.com/vncntcltt/airport-map
> 
> Broken atm (https://vncntcltt.github.io/airport-map/)
> 
> ```
> Failed to load resource: the server responded with a status of 404 ()
> Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0
> ```

Yes sorry it's fixed.

**Comment by mejrs** [comment]
I've done a lot of work with leaflet that requires async loading, what in particular would y'all want in an async tutorial?
 I'm thinking:

* extending _createTile(coords,done)_ to load using fetched json data
* importing google sheets, parsing its data and displaying it on a map
* using webassembly modules for computationally expensive calculations (e.g. a pathfinder)

Any suggestions?

**Comment by Malvoz** [comment]
> - [ ] Popular plugins (e.g. markercluster)

Adding [fullscreen control](https://github.com/Leaflet/Leaflet.fullscreen) to that item would be nice, since the [majority of internet users are on mobile](https://gs.statcounter.com/platform-market-share/desktop-mobile-tablet). Perhaps it could be part of https://leafletjs.com/examples/mobile/.

**Comment by Falke-Design** [comment]
Adding a tutorial for `PosAnimation`

**Comment by Malvoz** [comment]
- [ ] A security guide that could explain one or more of the following, if needed:

  - Using [CSP with Leaflet](https://github.com/Leaflet/Leaflet/issues?q=is%3Aissue+csp+is%3Aclosed) ([similar to MapBox](https://docs.mapbox.com/mapbox-gl-js/guides/browsers-and-testing/#csp-directives)). A [_strict_ CSP](https://web.dev/strict-csp/) is very useful for mitigating XSS ([#6465](https://github.com/Leaflet/Leaflet/issues/6465), [#7580](https://github.com/Leaflet/Leaflet/issues/7580))
  - [`crossOrigin`](https://leafletjs.com/reference.html#tilelayer-crossorigin)
  - [`referrerPolicy`](https://leafletjs.com/reference.html#tilelayer-referrerpolicy)
  - In the future perhaps, HTTP headers for Authorization: [#7560](https://github.com/Leaflet/Leaflet/issues/7560)

**Comment by kalvakeerthi** [comment]
I would like to work on this issue. Can you please assign this to me?

**Comment by Falke-Design** [comment]
@kalvakeerthi you can just make a PR , the issue is not needed to assigned to you.
What tutorials do you want to add?
